packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -6,372 +6,194 @@
+            evidence.blockHash == evidence.parentHash ||
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.signalRoot) <= 1 ||
+            // prover must not be zero
+            evidence.prover == address(0)
+        ) revert L1_INVALID_EVIDENCE();
+
+        TaikoData.ProposedBlock storage blk = state.proposedBlocks[
+            meta.id % config.maxNumProposedBlocks
+        ];
 
-        // ZK-prove block and mark block proven as invalid.
-        _proveBlock({
-            state: state,
-            config: config,
-            resolver: resolver,
-            proofVerifier: proofVerifier,
-            evidence: evidence,
-            target: target,
-            blockHashOverride: LibUtils.BLOCK_DEADEND_HASH
-        });
-    }
+        if (blk.metaHash != LibUtils.hashMetadata(meta))
+            revert L1_EVIDENCE_MISMATCH();
 
-    function _proveBlock(
-        TaikoData.State storage state,
-        TaikoData.Config memory config,
-        AddressResolver resolver,
-        IProofVerifier proofVerifier,
-        TaikoData.Evidence memory evidence,
-        TaikoData.BlockMetadata memory target,
-        bytes32 blockHashOverride
-    ) private {
-        if (evidence.meta.id != target.id) revert L1_ID();
-        if (evidence.prover == address(0)) revert L1_PROVER();
+        uint256 fcId = state.forkChoiceIds[blockId][evidence.parentHash];
+        if (fcId == 0) {
+            fcId = blk.nextForkChoiceId;
+            state.forkChoiceIds[blockId][evidence.parentHash] = fcId;
 
-        if (FLAG_CHECK_METADATA) {
-            if (
-                target.id <= state.latestVerifiedId ||
-                target.id >= state.nextBlockId
-            ) revert L1_ID();
-            if (
-                state
-                    .getProposedBlock(config.maxNumBlocks, target.id)
-                    .metaHash != target.hashMetadata()
-            ) revert L1_META_MISMATCH();
-        }
-
-        if (FLAG_VALIDATE_HEADER_FOR_METADATA) {
-            if (
-                evidence.header.parentHash == 0 ||
-                evidence.header.beneficiary != evidence.meta.beneficiary ||
-                evidence.header.difficulty != 0 ||
-                evidence.header.gasLimit !=
-                evidence.meta.gasLimit + config.anchorTxGasLimit ||
-                evidence.header.gasUsed == 0 ||
-                evidence.header.timestamp != evidence.meta.timestamp ||
-                evidence.header.extraData.length !=
-                evidence.meta.extraData.length ||
-                keccak256(evidence.header.extraData) !=
-                keccak256(evidence.meta.extraData) ||
-                evidence.header.mixHash != evidence.meta.mixHash
-            ) revert L1_META_MISMATCH();
+            unchecked {
+                ++blk.nextForkChoiceId;
+            }
+        } else if (fcId >= blk.nextForkChoiceId) {
+            revert L1_UNEXPECTED_FORK_CHOICE_ID(); // this shall not happen
```

Comment:
If i read the flow - how i see (and that might be totally wrong) - this revert can only be reached in 1 scenario:
1. Proof submitted and `maxVerificationsPerTx > 0` (meaninig the the prover will / might verify in the same TXN). This way the `blk.nextForkChoiceId` will be re-setted to 1 in `verifyBlock()`.
2. Then, some other prover wants to prove the same proposed block (tho is already proven and verified) with `proveBlock() `but this time the fcId will be = 1 while the the blk.nextForkChoiceId will also be 1.

So basically this line says: **'Hey Prover, this proposed block is already proven and verified !'**

Am I getting this right ? Also what i don't see is a scenario, where the fcId could be bigger then blk.nextForkChoiceId. (Only equal, in the above mentioned situation).

Reply:
I think everything you said is correct.